### PR TITLE
feat: Replace gorilla/mux router with echo

### LINKED ISSUE #556: Replacement for Gorilla/MUX package
# 🚀 Feature Request

### Relevant Package [**REQUIRED**]

 This feature request is for all of the EdgeX micro services have a dependency on it

### Description [**REQUIRED**]

This is a spin-off from https://github.com/edgexfoundry/edgex-go/issues/4247

The [gorilla/mux](https://github.com/gorilla/mux) package has been archived, and is no longer under active maintenance. We use it as the HTTP router in `go-mod-bootstrap`, `edgex-go` and other edgex services to define the http REST routes.

### Describe the solution you'd like
 
Per discussion in https://github.com/edgexfoundry/edgex-go/issues/4247, it will be replaced with Echo, which supports routing, middleware and tls features similar to `gorilla/mux`.

### Describe alternatives you've considered
 Have you considered any alternative solutions or workarounds?
Other alternatives were discussed in https://github.com/edgexfoundry/edgex-go/issues/4247, and Echo was the most appropriate solution in the end.

### PR DESCRIPTION
**Pull request by lindseysimple**
Replace gorilla/mux router with echo and modify the middleware structure to comply with it.

Closes #556.

If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-bootstrap/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [ ] I am not introducing a new dependency (add notes below if you are)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

### CODE REVIEW

File: bootstrap/handlers/httpcors.go [hunk 1]
```diff
@@ -46,27 +50,35 @@ func ProcessCORS(corsInfo config.CORSConfigurationInfo) func(http.Handler) http.
...
-	return func(w http.ResponseWriter, r *http.Request) {
-		if corsInfo.EnableCORS && r.Header.Get(Origin) != "" {
-			if len(corsInfo.CORSAllowedMethods) > 0 {
-				w.Header().Set(AccessControlAllowMethods, corsInfo.CORSAllowedMethods)
+func HandlePreflight(corsInfo config.CORSConfigurationInfo) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			r := c.Request()
+			// skip the middleware if AccessControlRequestMethod header is not defined
+			if r.Header.Get(AccessControlRequestMethod) == "" {
```
Comment (lenny-intel) [inline bootstrap/handlers/httpcors.go:64]: How would `AccessControlRequestMethod` be set?
Is this a bug in 3.0/2.x ?

Comment (lenny-intel) [reply bootstrap/handlers/httpcors.go:64]: @bnevis-i , so this change is also fixing a bug in 3.0 & 2.x, which should be documented.

Comment (bnevis-i) [reply bootstrap/handlers/httpcors.go:64]: No bug.  It's there in the previous release, but in httpserver.go

```
        b.router.Methods(http.MethodOptions).MatcherFunc(func(r *http.Request, rm *mux.RouteMatch) bool {
                return r.Header.Get(AccessControlRequestMethod) != ""
        }).HandlerFunc(HandlePreflight(bootstrapConfig.Service.CORSConfiguration))
```

**Review by lenny-intel — APPROVED** [review]
LGTM

**Review by bnevis-i — APPROVED** [review]
LGTM as well.  Feels a lot like Express JS.

**Review by lenny-intel — CHANGES_REQUESTED** [review]
This PR now needs to include the new Common Controller code added by this PR: https://github.com/edgexfoundry/go-mod-bootstrap/pull/562

**Review by lenny-intel — APPROVED** [review]
LGTM

File: bootstrap/handlers/auth_middleware.go [hunk 2]
```diff
@@ -54,28 +60,32 @@ func VaultAuthenticationHandlerFunc(secretProvider interfaces.SecretProviderExt,
...
-				return
+				return inner(c)
 			}
-			lc.Errorf("Unable to parse JWT for call to '%s'; unauthorized", r.URL.Path)
-			http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+			err := fmt.Errorf("unable to parse JWT for call to '%s'; unauthorized", r.URL.Path)
+			lc.Errorf("%v", err)
+			// set Response.Committed to true in order to rewrite the status code
+			w.Committed = false
+			return echo.NewHTTPError(http.StatusUnauthorized, err.Error())
```
Comment (bnevis-i) [inline bootstrap/handlers/auth_middleware.go:79]: For the access denied case, let's not return too much information to the caller as to why the access was denied.

Comment (lindseysimple) [reply bootstrap/handlers/auth_middleware.go]: Thanks @bnevis-i, response message updated as you suggested.

File: bootstrap/handlers/auth_middleware.go [hunk 3]
```diff
@@ -54,28 +60,32 @@ func VaultAuthenticationHandlerFunc(secretProvider interfaces.SecretProviderExt,
 				validToken, err := secretProvider.IsJWTValid(token)
 				if err != nil {
 					lc.Errorf("Error checking JWT validity: %v", err)
-					http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
-					return
+					// set Response.Committed to true in order to rewrite the status code
+					w.Committed = false
+					return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
```
Comment (bnevis-i) [inline bootstrap/handlers/auth_middleware.go:65]: This one is still potentially leaking a lot of information.

File: bootstrap/handlers/auth_middleware.go (hunk 2 above)
Comment (bnevis-i) [reply bootstrap/handlers/auth_middleware.go]: @lindseysimple  There was another one earlier that was also putting out err.Error().  I also added a comment there as well.

File: bootstrap/handlers/auth_middleware.go (hunk 3 above)
Comment (lindseysimple) [reply bootstrap/handlers/auth_middleware.go]: Thanks @bnevis-i for catching this place as well! Change updated.

**Pull request merged**